infrastructures/infrastructure-azure/src/main/java/org/ow2/proactive/resourcemanager/nodesource/billing/AzureBillingResourceUsage.java

```diff
@@ -184,17 +186,27 @@ String getLastResourceUsageHistory(AzureBillingCredentials azureBillingCredentia
 
             } else if (jsonObject.has("error")) { // HISTORY NOT RETRIEVED BUT TRY AGAIN !!
 
-                String queryErrorCodeMessage = jsonObject.get("error").getAsJsonObject().get("code").getAsString();
+                JsonObject queryErrorObject = jsonObject.get("error").getAsJsonObject();
+                String queryErrorCode = queryErrorObject.get("code").getAsString();
+                String queryErrorMessage = queryErrorObject.get("message").getAsString();
 
-                if (queryErrorCodeMessage.equals("ExpiredAuthenticationToken")) {
-                    LOGGER.debug("AzureBillingResourceUsage getLastResourceUsageHistory ExpiredAuthenticationToken");
+                if (queryErrorCode.equals("ExpiredAuthenticationToken")) {
+                    LOGGER.debug("AzureBillingResourceUsage " + queryErrorCode + ":" + queryErrorMessage +
+                                 "(renewing token)");
                     azureBillingCredentials.renewOrOnlyGetAccessToken(true);
                     continue;
-                } else if (queryErrorCodeMessage.equals("ProcessingNotCompleted")) {
+                } else if (queryErrorCode.equals("ProcessingNotCompleted") ||
+                           (queryErrorCode.equals("InvalidInput") &&
+                            queryErrorMessage.equals("reportedendtime cannot be in the future."))) {
                     endDateTime = endDateTime.minusHours(1);
-                    LOGGER.debug("AzureBillingResourceUsage getLastResourceUsageHistory ProcessingNotCompleted new endDateTime " +
-                                 endDateTime);
+                    LOGGER.debug("AzureBillingResourceUsage " + queryErrorCode + ":" + queryErrorMessage +
+                                 "(new endDateTime=" + endDateTime + ")");
                     continue;
+                } else if (queryErrorCode.equals("InvalidInput") &&
+                           queryErrorMessage.equals("reportedstarttime cannot be in the future.")) {
+                    LOGGER.debug("AzureBillingResourceUsage " + queryErrorCode + ":" + queryErrorMessage +
+                                 "(let's try next time with the same reportedstarttime)");
+                    return null;
```

Comment:
Is it the place to throw an exception instead of sending a null value ? `AzureBillingException` could provide the right semantic here.

Reply:
At this method call level, either AzureBillingException is catched and cost getter threads  shutdowns, or a null is returned and nothing is done. We decide to shutdown threads if an unexpected error is returned when querying the Azure server.